torch/nn/parallel/distributed.py

```diff
@@ -117,33 +123,64 @@ def __init__(self, module, device_ids=None, output_device=None, dim=0):
         else:
             self._module_copies = [self.module]
 
-        # Split parameters into buckets that will coalesce reductions
-        # TODO: different types need different buckets
-        t = None
-        for p in self.module.parameters():
-            tp = type(p.data)
-            if t is not None and t is not tp:
-                raise ValueError("DistributedDataParallel requires all parameters' data to be of the same type")
-            t = tp
+        # Split parameters into type buckets so that parameter sync (broadcast)
+        # can operates on mixed parameter types. (e.g. mixed half and float)
+        self.param_type_buckets = \
+            defaultdict(lambda: [[] for _ in range(len(self.device_ids))])
 
+        for device_idx, module in enumerate(self._module_copies):
+            for p in module.parameters():
+                tp = p.type()
+                if tp == torch.cuda.HalfTensor and \
+                        dist._backend != dist.dist_backend.NCCL and \
+                        dist._backend != dist.dist_backend.GLOO:
+                    raise RuntimeError("DistributedDataParallel currently only "
+                                       "supports half precision parameters "
+                                       "with NCCL backend")
+                # Add the parameter into the type bucket
+                self.param_type_buckets[tp][device_idx].append(p)
+
+        # TODO, adding mixed precision support in NCCL reduction code path
+        # This is because NCCL backend doesn't support multiple reduction
+        # bucket
```

Comment:
@apaszke This will be further added in https://github.com/pytorch/pytorch/pull/5064